src/libraries/System.Runtime/ref/System.Runtime.cs

```diff
@@ -968,6 +968,11 @@ public static partial class Convert
         public static double ToDouble(uint value) { throw null; }
         [System.CLSCompliantAttribute(false)]
         public static double ToDouble(ulong value) { throw null; }
+        public static byte[] FromHexString(string s) { throw null; }
+        public static byte[] FromHexString(System.ReadOnlySpan<char> chars) { throw null; }
+        public static string ToHexString(byte[] inArray) { throw null; }
+        public static string ToHexString(byte[] inArray, int offset, int length) { throw null; }
+        public static string ToHexString(System.ReadOnlySpan<byte> bytes) { throw null; }
```

Comment:
Are there any places in the dotnet/runtime libraries where we're currently doing these operations manually such that they could instead be switched to call these APIs?

Reply:
* Most `byte -> string` conversions have already consolidated to use the internal HexConverter as part of #1273. Some places could however profit from the new APIs (e.g. src\libraries\Common\tests\System\Security\Cryptography\ByteUtils.cs)
* There are a few `string -> byte` conversions across the libraries. Some (e.g. ActivityTraceId.HexDigitToBinary) are, however, case sensitive...

Would you like to see a cleanup/consolidation as part of this PR or as a follow-up?

Reply:
My question stems more from wanting to make sure the methods are actually useful (that's a question about the approved API design rather than about your work in the PR).  If with all of the places we do hex conversions we don't actually have any places that would benefit from these APIs, it makes me question whether they're the right APIs to be adding.

cc: @bartonjs, @GrabYourPitchforks

Reply:
Well, my ByteArrayToHex() methods could certainly be replaced with this... if all of my compilation targets were net5+ (I don't remember if I have tests that run on netfx anymore).

There's some hex parsing and hex producing code in the crypto libraries themselves... the parser ignores any non-hex data (at this point, for compat; historically, because... uh... clearly it was whitespace?) so probably can't change.  The hex producers, possibly... if they're only used in net5 targets (the Pkcs and Xml libraries are netstandard OOBs, but I feel like the hex producing code is in S.S.C.Encoding, which is in the framework, so it might be able to change, unless it does weird things with whitespace).

Reply:
@stephentoub An analysis of the usefulness of the new API for use within the dotnet/runtime libraries

Could profit from new API (ignoring potential issues with compilation targets):
```
src\libraries\Common\tests\System\Security\Cryptography\ByteUtils.cs:25
src\libraries\Common\tests\System\Security\Cryptography\ByteUtils.cs:53
src\libraries\System.DirectoryServices.AccountManagement\src\System\DirectoryServices\AccountManagement\Utils.cs:87
src\libraries\System.Drawing.Common\tests\mono\System.Drawing\BitmapTests.cs:393
src\libraries\System.Net.NetworkInformation\src\System\Net\NetworkInformation\PhysicalAddress.cs:93
src\libraries\System.Reflection.MetadataLoadContext\tests\src\TestUtils\TestUtils.cs:108
src\libraries\System.Reflection.MetadataLoadContext\tests\src\TestUtils\TestUtils.cs:121
src\libraries\System.Runtime\tests\System\Text\Unicode\Utf8Tests.cs:64
src\libraries\System.Security.Cryptography.Encoding\src\Internal\Cryptography\AsnFormatter.cs:15
src\libraries\System.Security.Cryptography.X509Certificates\src\Internal\Cryptography\Helpers.cs:36
src\libraries\System.Windows.Extensions\tests\X509Certificate2UIManualTests.cs:73
```

Could profit from making access to the lookup table in `HexConverter` public:
```
src\libraries\System.Security.Cryptography.X509Certificates\src\Internal\Cryptography\Helpers.cs:98
```

Cannot profit from the new methods due to limitations of the new API
```
src\libraries\Common\tests\System\Net\Http\LoopbackServer.AuthenticationHelpers.cs:287 - Encode lowercase
src\libraries\System.Diagnostics.DiagnosticSource\src\System\Diagnostics\Activity.cs - Encode lowercase
src\libraries\System.DirectoryServices.AccountManagement\src\System\DirectoryServices\AccountManagement\Utils.cs:75 - Encode lowercase
src\libraries\System.Runtime.Numerics\tests\BigInteger\MyBigInt.cs:903 - Decode lowercase, space-separated
src\libraries\System.Security.Cryptography.Encoding\src\Internal\Cryptography\AsnFormatter.cs:22 - Encode space-separated
src\libraries\System.Security.Cryptography.Pkcs\src\Internal\Cryptography\PkcsHelpers.cs:348 - Decode (enforce uppercase)
src\libraries\System.Security.Cryptography.X509Certificates\src\Internal\Cryptography\Helpers.cs:24 - Decode to Span<char>
src\libraries\System.Security.Cryptography.X509Certificates\src\Internal\Cryptography\Helpers.cs:41 - Decode white space
```

It could be that I missed a few additional places...

Reply:
Thanks for the analysis!

@bartonjs, @GrabYourPitchforks, we still believe the APIs approved and implemented publicly in this PR are the right ones?

Personally, I'd want to see at least at least a couple of existing call sites switched in this PR, or else I question the real worthwhileness of the particular shapes employed here.  We already have BitConverter.ToString which outputs at least one pattern; this (at least the ToString direction) is just outputting another form.  I question whether this form is so much more useful that it's worth adding.

Reply:
> we still believe the APIs approved and implemented publicly in this PR are the right ones?

I don't remember why we didn't also do the span-writing versions, but these are definitely better than (leads into)

>  We already have BitConverter.ToString which outputs at least one pattern; this (at least the ToString direction) is just outputting another form.

While I use that one in a pinch, I generally don't want the hyphens.  The [top hit on StackOverflow for "BitConverter.ToString"](https://stackoverflow.com/search?q=bitconverter.tostring) is

```C#
public static string ByteArrayToString(byte[] ba) { return BitConverter.ToString(ba).Replace("-",""); } 
```

Suggesting that the form here is what many (if not most) callers want.  So I feel there's definite goodness.